x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/conditional/Intersects.java

```diff
@@ -0,0 +1,234 @@
+                    } else {
+                        // Left is a constant string, right is a variable string
+                        return () -> new IntersectsConstantStringEvaluator(asString(left()), toEvaluator.apply(right()).get());
+                    }
+                } else if (right().foldable()) {
+                    // Left is a variable string, and right is a constant string
+                    return () -> new IntersectsConstantStringEvaluator(asString(right()), toEvaluator.apply(left()).get());
+                } else {
+                    // Both are variable strings
+                    return () -> new IntersectsStringStringEvaluator(toEvaluator.apply(left()).get(), toEvaluator.apply(right()).get());
+                }
+            } else if (left().foldable()) {
+                // Left is constant string, right is long
+                return () -> new IntersectsLongConstantEvaluator(toEvaluator.apply(right()).get(), asString(left()));
+            } else {
+                // Left is variable string, right is long
+                return () -> new IntersectsLongStringEvaluator(toEvaluator.apply(right()).get(), toEvaluator.apply(left()).get());
+            }
+        } else if (isString(right().dataType())) {
+            if (right().foldable()) {
+                // Left is long, right is constant string
+                return () -> new IntersectsLongConstantEvaluator(toEvaluator.apply(left()).get(), asString(right()));
+            } else {
+                // Left is long, right is variable string
+                return () -> new IntersectsLongStringEvaluator(toEvaluator.apply(left()).get(), toEvaluator.apply(right()).get());
+            }
+        } else {
+            // left and right are longs
+            return () -> new IntersectsLongLongEvaluator(toEvaluator.apply(left()).get(), toEvaluator.apply(right()).get());
+        }
+    }
+
+    private static String asString(Expression expression) {
+        Object result = expression.fold();
+        if (result instanceof BytesRef bytesRef) {
+            return bytesRef.utf8ToString();
+        } else {
+            // TODO: Should this be an error case?
+            return result.toString();
+        }
+    }
+
+    @Evaluator(extraName = "LongLong")
+    static boolean process(long leftValue, long rightValue) {
+        return leftValue == rightValue;
+    }
+
+    @Evaluator(extraName = "LongConstant")
+    static boolean processConstant(long leftValue, @Fixed String rightValue) {
+        Geometry geometry = SpatialUtils.stringAsGeometry(rightValue);
+        return pointIntersectsGeometry(leftValue, geometry);
+    }
+
+    @Evaluator(extraName = "LongString")
+    static boolean process(long leftValue, BytesRef rightValue) {
+        Geometry geometry = SpatialUtils.stringAsGeometry(rightValue.utf8ToString());
+        return pointIntersectsGeometry(leftValue, geometry);
```

Comment:
I think this'll happen when you try to compare a geo_point field to a keyword field. Like, a random one. I think *that* should be a compile error on the query and we should make people do `INTERSECTS(geo_point, TO_GEOPOINT(kwd))`.

Reply:
OK, so reduce the scope of this function to one geo_point and one constant only. Sounds sensible. I'll leave the PR as is, but we do this for the production version.